File: backend/app/schemas/place.py
from enum import Enum
from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel, Field


class PlaceSource(str, Enum):
    AMAP_POI = "amap_poi"
    RAG = "rag"
    SYNTHESIZED = "synthesized"


class PlaceCategory(str, Enum):
    ATTRACTION = "attraction"
    FOOD = "food"
    HOTEL = "hotel"
    TRANSPORT = "transport"
    UNKNOWN = "unknown"


class RetrievalExecutionMode(str, Enum):
    """How a provider-backed place was obtained for this request."""

    LIVE = "live"
    FIXTURE = "fixture"
    FALLBACK = "fallback"


class EvidenceStatus(str, Enum):
    """Evidence boundary for a user-requested place attribute."""

    VERIFIED = "VERIFIED"
    UNKNOWN = "UNKNOWN"
    REQUIRES_CONFIRMATION = "REQUIRES_CONFIRMATION"


class Coordinates(BaseModel):
    lng: float
    lat: float


class PlaceRAGMeta(BaseModel):
    tip_snippets: list[str] = Field(default_factory=list, description="从游记提取的避坑/推荐语，最多3条")
    sentiment_score: float = Field(default=0.0, ge=-1.0, le=1.0, description="游记情感倾向 -1~1")
    source_note_ids: list[str] = Field(default_factory=list, description="支撑该内容的游记文档 ID（可溯源）")


class ConstraintEvidence(BaseModel):
    constraint: str = Field(description="稳定的约束键，例如 family_room")
    label: str = Field(description="面向用户的约束名称")
    status: EvidenceStatus
    detail: str = Field(description="证据边界或下一步确认说明")
    source: str = Field(description="证据来源，例如 amap_poi / unavailable_in_poi")
    value: Optional[Any] = Field(None, description="用于判断的结构化原值")
    source_url: Optional[str] = Field(None, description="可打开的来源入口；提供方无字段级 URL 时为空")
    observed_at: Optional[datetime] = Field(None, description="本次来源观测时间")
    confidence: float = Field(default=0.0, ge=0, le=1)


class GeoEvidence(BaseModel):
    slot_id: str
    anchor_place: str
    constraint_kind: str = Field(
        default="proximity",
        description="proximity | route；半径和路线证据分开，避免一个 UNKNOWN 覆盖已验证坐标",
    )
    status: EvidenceStatus
    satisfies_constraint: Optional[bool] = None
    straight_line_distance_km: Optional[float] = None
    estimated_travel_minutes: Optional[int] = None
    transfer_count: Optional[int] = None
    transport_mode: str = "walking"
    source: str = "amap_coordinates"
    route_response_hash: Optional[str] = None
    observed_at: Optional[datetime] = None
    failure_reason: Optional[str] = None


class Place(BaseModel):
    """标准化地点对象 - 系统全局货币，贯穿 LangGraph 三个节点和前端状态树"""
    place_id: str = Field(..., description="高德 POI ID，全局唯一")
    name: str
    category: PlaceCategory
    address: str
    coords: Coordinates
    city: str
    district: Optional[str] = None
    source: PlaceSource = PlaceSource.SYNTHESIZED
    execution_mode: Optional[RetrievalExecutionMode] = Field(
        None,
        description="本次检索执行模式；与内容来源 source 分开，禁止用 amap_poi 冒充 live",
    )
    retrieval_provider: Optional[str] = Field(None, description="本次检索提供方，例如 amap")
    retrieval_request_hash: Optional[str] = Field(None, description="脱敏后提供方请求参数的 SHA256")
    retrieval_response_hash: Optional[str] = Field(None, description="提供方响应的 SHA256")
    retrieval_observed_at: Optional[datetime] = Field(None, description="本次提供方响应观测时间")
    recommendation_slot_ids: list[str] = Field(
        default_factory=list,
        description="该候选满足的 RecommendationPlan 槽位；实体去重时取并集",
    )
    canonical_entity_names: list[str] = Field(
        default_factory=list,
        description="由显式实体槽位绑定的 canonical 名称，不由 LLM 或 POI 名称猜测",
    )

    # 高德客观数据
    amap_rating: Optional[float] = Field(None, ge=0, le=5, description="高德评分 0-5")
    amap_price: Optional[float] = Field(None, description="人均消费（元）")
    opening_hours: Optional[str] = None
    phone: Optional[str] = None
    amap_photos: list[str] = Field(default_factory=list, description="高德图片 URL 列表")

    # RAG 主观数据（无游记命中则为 None）
    rag_meta: Optional[PlaceRAGMeta] = None

    # AI 生成的描述信息
    description: Optional[str] = Field(None, description="一句话特点描述，20-40字")
    tags: list[str] = Field(default_factory=list, description="适合人群/场景标签，如 ['情侣', '拍照', '亲子']")
    constraint_evidence: list[ConstraintEvidence] = Field(
        default_factory=list,
        description="本轮用户约束在该地点上的结构化证据状态；UNKNOWN 不等于满足",
    )
    selection_evidence_status: Optional[EvidenceStatus] = Field(
        None,
        description="SelectionPolicy 汇总后的候选证据等级；UNKNOWN/REQUIRES_CONFIRMATION 不视为满足高风险约束",
    )
    geo_evidence: list[GeoEvidence] = Field(
        default_factory=list,
        description="相对 RecommendationPlan 锚点的坐标距离与预计通勤证据",
    )
    confirmation_actions: list[str] = Field(
        default_factory=list,
        description="对 REQUIRES_CONFIRMATION 约束给出可执行联系动作，不把确认动作冒充已满足",
    )

    # Optimizer 节点写入
    cluster_id: Optional[int] = Field(None, description="K-Means 分配的日期簇 ID")
    visit_order: Optional[int] = Field(None, description="簇内 TSP 排序序号")
    estimated_duration: Optional[int] = Field(None, description="建议游览时长（分钟）")
    duration_basis: Optional[str] = Field(None, description="时长来源：llm/rule/user")

    # 温馨提示（TipsGenerator 写入）
    tips: list[str] = Field(default_factory=list, description="出行温馨提示，如取号/招牌菜/天气")

File: frontend/src/types/place.ts
export type PlaceSource = 'amap_poi' | 'rag' | 'synthesized'
export type PlaceCategory = 'attraction' | 'food' | 'hotel' | 'transport'

export interface Coordinates {
  lng: number
  lat: number
}

export interface PlaceRAGMeta {
  tipSnippets: string[]       // 从游记提取的避坑/推荐语，最多3条
  sentimentScore: number      // -1 ~ 1
  sourceNoteIds: string[]     // 支撑该内容的游记文档 ID（可溯源）
}

export type EvidenceStatus = 'VERIFIED' | 'UNKNOWN' | 'REQUIRES_CONFIRMATION'

export interface ConstraintEvidence {
  constraint: string
  label: string
  status: EvidenceStatus
  detail: string
  source: string
  value?: unknown
  sourceUrl?: string
  observedAt?: string
  confidence: number
}

export interface GeoEvidence {
  slotId: string
  anchorPlace: string
  status: EvidenceStatus
  satisfiesConstraint?: boolean
  straightLineDistanceKm?: number
  estimatedTravelMinutes?: number
  transportMode: string
  source: string
}

/** Phase B：替代方案 */
export interface PlaceAlternative {
  placeId: string
  name: string
  whyAlternative: string      // "比 A 更便宜 / 排队少 / 更适合带娃"
}

/** Phase B：结构化推荐（地点卡背面展示） */
export interface PlaceRecommendation {
  placeId: string
  name: string
  categoryL1: string
  categoryL2: string
  reason: string              // 推荐理由（引自游记，可为空）
  suitableFor: string[]       // 适合人群
  avoidTips: string[]         // 避坑提示
  sourceChunkIds: string[]    // 游记 chunk 来源
  alternatives: PlaceAlternative[]
  confidence: 'high' | 'medium' | 'low'
}

export interface Place {
  placeId: string
  name: string
  category: PlaceCategory
  address: string
  coords: Coordinates
  city: string
  district?: string
  source: PlaceSource

  // 高德客观数据
  amapRating?: number
  amapPrice?: number
  openingHours?: string
  phone?: string
  amapPhotos: string[]

  // RAG 主观数据
  ragMeta?: PlaceRAGMeta

  // AI 生成
  description?: string
  tags: string[]
  constraintEvidence: ConstraintEvidence[]
  selectionEvidenceStatus?: EvidenceStatus
  geoEvidence: GeoEvidence[]
  confirmationActions: string[]

  // Optimizer
  clusterId?: number
  visitOrder?: number
  estimatedDuration?: number

  // Phase B：结构化推荐（可选，由 Synthesizer v2 填入）
  recommendation?: PlaceRecommendation
}

/** 将后端蛇形命名转换为前端驼峰命名 */
export function parsePlaceFromAPI(raw: Record<string, unknown>): Place {
  const rec = raw.recommendation as Record<string, unknown> | undefined
  return {
    placeId: raw.place_id as string,
    name: raw.name as string,
    category: raw.category as PlaceCategory,
    address: raw.address as string,
    coords: raw.coords as Coordinates,
    city: raw.city as string,
    district: raw.district as string | undefined,
    source: (raw.source as PlaceSource) || 'synthesized',
    amapRating: raw.amap_rating as number | undefined,
    amapPrice: raw.amap_price as number | undefined,
    openingHours: raw.opening_hours as string | undefined,
    phone: raw.phone as string | undefined,
    amapPhotos: (raw.amap_photos as string[]) || [],
    ragMeta: raw.rag_meta
      ? {
          tipSnippets: (raw.rag_meta as Record<string, unknown>).tip_snippets as string[],
          sentimentScore: (raw.rag_meta as Record<string, unknown>).sentiment_score as number,
          sourceNoteIds: (raw.rag_meta as Record<string, unknown>).source_note_ids as string[],
        }
      : undefined,
    description: raw.description as string | undefined,
    tags: (raw.tags as string[]) || [],
    constraintEvidence: ((raw.constraint_evidence as Record<string, unknown>[]) || []).map((item) => ({
      constraint: item.constraint as string,
      label: item.label as string,
      status: item.status as EvidenceStatus,
      detail: item.detail as string,
      source: item.source as string,
      value: item.value,
      sourceUrl: item.source_url as string | undefined,
      observedAt: item.observed_at as string | undefined,
      confidence: Number(item.confidence || 0),
    })),
    selectionEvidenceStatus: raw.selection_evidence_status as EvidenceStatus | undefined,
    geoEvidence: ((raw.geo_evidence as Record<string, unknown>[]) || []).map((item) => ({
      slotId: item.slot_id as string,
      anchorPlace: item.anchor_place as string,
      status: item.status as EvidenceStatus,
      satisfiesConstraint: item.satisfies_constraint as boolean | undefined,
      straightLineDistanceKm: item.straight_line_distance_km as number | undefined,
      estimatedTravelMinutes: item.estimated_travel_minutes as number | undefined,
      transportMode: String(item.transport_mode || 'walking'),
      source: String(item.source || 'amap_coordinates'),
    })),
    confirmationActions: (raw.confirmation_actions as string[]) || [],
    clusterId: raw.cluster_id as number | undefined,
    visitOrder: raw.visit_order as number | undefined,
    estimatedDuration: raw.estimated_duration as number | undefined,
    recommendation: rec
      ? {
          placeId: rec.place_id as string,
          name: rec.name as string,
          categoryL1: (rec.category_l1 as string) || '',
          categoryL2: (rec.category_l2 as string) || '',
          reason: (rec.reason as string) || '',
          suitableFor: (rec.suitable_for as string[]) || [],
          avoidTips: (rec.avoid_tips as string[]) || [],
          sourceChunkIds: (rec.source_chunk_ids as string[]) || [],
          alternatives: ((rec.alternatives as Record<string, unknown>[]) || []).map((a) => ({
            placeId: a.place_id as string,
            name: a.name as string,
            whyAlternative: (a.why_alternative as string) || '',
          })),
          confidence: (rec.confidence as 'high' | 'medium' | 'low') || 'low',
        }
      : undefined,
  }
}
